eclass/rocm.eclass

```diff
@@ -0,0 +1,232 @@
+# Copyright 2022 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+# @ECLASS: rocm.eclass
+# @MAINTAINER:
+# Gentoo Science Project <sci@gentoo.org>
+# @AUTHOR:
+# Yiyang Wu <xgreenlandforwyy@gmail.com>
+# @SUPPORTED_EAPIS: 7 8
+# @BLURB: Common functions and variables for ROCm packages written in HIP
+# @DESCRIPTION:
+# ROCm packages such as sci-libs/<roc|hip>*, and packages built on top of ROCm
+# libraries, can utilize variables and functions provided by this eclass.
+# Currently, it handles the AMDGPU_TARGETS variable via USE_EXPAND, so user can
+# edit USE flag to control which GPU architecture to compile. Using
+# ${ROCM_USEDEP} can ensure coherence among dependencies. Ebuilds can call the
+# function get_amdgpu_flag to translate activated target to GPU compile flags,
+# passing it to configuration. Function check_amdgpu can help ebuild ensure
+# read and write permissions to GPU device in src_test phase, throwing friendly
+# error message if permission denied.
+#
+# @EXAMPLE:
+# @CODE
+# # Example ebuild for ROCm library in https://github.com/ROCmSoftwarePlatform
+# # which uses cmake to build and test, and depends on rocBLAS:
+# inherit cmake rocm
+# # ROCm libraries SRC_URI is usually in form of:
+# SRC_URI="https://github.com/ROCmSoftwarePlatform/${PN}/archive/rocm-${PV}.tar.gz -> ${P}.tar.gz"
+# S=${WORKDIR}/${PN}-rocm-${PV}
+# SLOT="0/$(ver_cut 1-2)"
+# IUSE="test"
+# REQUIRED_USE="${ROCM_REQUIRED_USE}"
+# RESTRICT="!test? ( test )"
+#
+# RDEPEND="
+#     dev-util/hip
+#     sci-libs/rocBLAS:${SLOT}[${ROCM_USEDEP}]
+# "
+#
+# src_configure() {
+#     # avoid sandbox violation
+#     addpredict /dev/kfd
+#     addpredict /dev/dri/
+#     local mycmakeargs=(
+#         -DAMDGPU_TARGETS="$(get_amdgpu_flags)"
+#         -DBUILD_CLIENTS_TESTS=$(usex test ON OFF)
+#     )
+#     CXX=hipcc cmake_src_configure
+# }
+#
+# src_test() {
+#     # grant access to AMD GPU device
+#     for device in /dev/kfd /dev/dri/render*; do
+#         addwrite "${device}"
+#     done
+#     check_amdgpu
```

Comment:
I'm a bit confused… why is `addwrite` inline but `check_amdgpu` standarized? Is there ever a case for doing one and not the other?

Reply:
Hmm the complete history is:

At first `check_amdgpu` is `check_rw_permission`, checking permissions on one file. @heroxbd suggested me to narrowed down its usage, restricting to check specific devices (because the die message is all about render group, which is helpful for checking amdgpu devices, but not helpful in other occasions).

So I changed `check_rw_permission <device>`  to `check_amdgpu`

I think it can go back: only change the name `check_rw_permission` to `check_amdgpu_device`, then in ebuild:
    for device in /dev/kfd /dev/dri/render*; do
        addwrite "${device}"
        check_amdgpu_device "${device}"
    done
```

Reply:
Maybe I misunderstood your point previously. Now I put `addwrite` to check_amdgpu, so there is no loop on devices in the ebuild `src_test`, which seems much cleaner.